api/stacks/deployments/deployer_remote.go

```diff
@@ -0,0 +1,282 @@
+
+	return d.remoteStack(stack, endpoint, unpackerCmdBuilderOptions{
+		stackOperation: OperationSwarmDeploy,
+		pullImage:      pullImage,
+		prune:          prune,
+		registries:     registries,
+	})
+}
+
+// Undeploy a swarm stack on remote environment using a https://github.com/portainer/compose-unpacker container
+func (d *stackDeployer) UndeployRemoteSwarmStack(stack *portainer.Stack, endpoint *portainer.Endpoint) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return d.remoteStack(stack, endpoint, unpackerCmdBuilderOptions{
+		stackOperation: OperationSwarmUndeploy,
+	})
+}
+
+// Start a swarm stack on remote environment using a https://github.com/portainer/compose-unpacker container
+func (d *stackDeployer) StartRemoteSwarmStack(stack *portainer.Stack, endpoint *portainer.Endpoint) error {
+	return d.remoteStack(stack, endpoint, unpackerCmdBuilderOptions{
+		stackOperation: OperationSwarmStart,
+	})
+}
+
+// Stop a swarm stack on remote environment using a https://github.com/portainer/compose-unpacker container
+func (d *stackDeployer) StopRemoteSwarmStack(stack *portainer.Stack, endpoint *portainer.Endpoint) error {
+	return d.remoteStack(stack, endpoint, unpackerCmdBuilderOptions{
+		stackOperation: OperationSwarmStop,
+	})
+}
+
+// Does all the heavy lifting:
+// * connect to env
+// * build the args for compose-unpacker
+// * deploy compose-unpacker container
+// * wait for deployment to end
+// * gather deployment logs and bubble them up
+func (d *stackDeployer) remoteStack(stack *portainer.Stack, endpoint *portainer.Endpoint, opts unpackerCmdBuilderOptions) error {
+	ctx := context.TODO()
+
+	cli, err := d.createDockerClient(ctx, endpoint)
+	if err != nil {
+		return errors.WithMessage(err, "unable to create docker client")
+	}
+	defer cli.Close()
+
+	image := getUnpackerImage()
+
+	reader, err := cli.ImagePull(ctx, image, types.ImagePullOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to pull unpacker image")
+	}
+	defer reader.Close()
+	io.Copy(io.Discard, reader)
+
+	info, err := cli.Info(ctx)
+	if err != nil {
+		return errors.Wrap(err, "unable to get agent info")
+	}
+	targetSocketBind := getTargetSocketBind(info.OSType)
+
+	composeDestination := filesystem.JoinPaths(stack.ProjectPath, composePathPrefix)
+
+	opts.composeDestination = composeDestination
+
+	cmd := d.buildUnpackerCmdForStack(stack, opts)
+
+	log.Debug().
+		Str("image", image).
+		Str("cmd", strings.Join(cmd, " ")).
+		Msg("running unpacker")
+
+	rand.Seed(time.Now().UnixNano())
+	resp, err := cli.ContainerCreate(ctx, &container.Config{
+		Image: image,
+		Cmd:   cmd,
+	}, &container.HostConfig{
+		Binds: []string{
+			fmt.Sprintf("%s:%s", composeDestination, composeDestination),
+			fmt.Sprintf("%s:%s", targetSocketBind, targetSocketBind),
+		},
+	}, nil, nil, fmt.Sprintf("portainer-unpacker-%d-%s-%d", stack.ID, stack.Name, rand.Intn(100)))
+
+	defer cli.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{})
```

Comment:
shouldn't this be after the err check? if `ContainerCreate` failed, then `resp` can be null so `resp.ID` will panic